### PR TITLE
WiP: MINIFICPP-839 - Fix memleaks

### PR DESCRIPTION
**Pull request by bakaid**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/src/core/repository/VolatileContentRepository.cpp [hunk 1]
```diff
@@ -81,10 +81,7 @@ void VolatileContentRepository::start() {
     return;
   if (running_)
     return;
-  thread_ = std::thread(&VolatileContentRepository::run, shared_from_parent<VolatileContentRepository>());
-  thread_.detach();
```
Comment (phrocker) [inline libminifi/src/core/repository/VolatileContentRepository.cpp]: What was the point of removing this?

File: libminifi/include/core/FlowFile.h [hunk 2]
```diff
@@ -254,29 +254,29 @@ class FlowFile : public core::Connectable, public ReferenceContainer {
    * Sets the original connection with a shared pointer.
    * @param connection shared connection.
```
Comment (phrocker) [inline libminifi/include/core/FlowFile.h]: Hmm, the flowfile should complete its activity. Why does this need to be weak?

File: libminifi/src/ResourceClaim.cpp [hunk 3]
```diff
@@ -41,11 +42,15 @@ void setDefaultDirectory(std::string path) {
   default_directory_path = path;
 }
 
-ResourceClaim::ResourceClaim(std::shared_ptr<core::StreamManager<ResourceClaim>> claim_manager)
```
Comment (phrocker) [inline libminifi/src/ResourceClaim.cpp]: These claims should not be persisted on stops. They should complete, so making them weak seems counterintuitive, though I admit it may be necessary -- why is the claim required to be weak ?

File: libminifi/src/core/ProcessSession.cpp [hunk 4]
```diff
@@ -805,7 +805,7 @@ void ProcessSession::commit() {
         continue;
       }
 
-      connection = std::static_pointer_cast<Connection>(record->getConnection());
+      connection = std::static_pointer_cast<Connection>(record->getConnection().lock());
```
Comment (phrocker) [inline libminifi/src/core/ProcessSession.cpp]: Session should be a hard pointer. How did you arrive at this?

File: libminifi/src/core/Processor.cpp [hunk 5]
```diff
@@ -176,7 +176,7 @@ void Processor::removeConnection(std::shared_ptr<Connectable> conn) {
     // Connection is destination to the current processor
     if (_incomingConnections.find(connection) != _incomingConnections.end()) {
       _incomingConnections.erase(connection);
-      connection->setDestination(NULL);
```
Comment (phrocker) [inline libminifi/src/core/Processor.cpp]: hmmm, why this change? I would prefer nullptr here. It's easier to read IMO, so I think we should either leave this as-is or use nullptr.

File: libminifi/include/core/FlowFile.h (hunk 2 above)
Comment (phrocker) [reply libminifi/include/core/FlowFile.h]: I presume this is because the processor is weak?

File: libminifi/include/ResourceClaim.h [hunk 6]
```diff
@@ -112,7 +122,7 @@ class ResourceClaim : public std::enable_shared_from_this<ResourceClaim> {
   // Full path to the content
   std::string _contentFullPath;
 
-  std::shared_ptr<core::StreamManager<ResourceClaim>> claim_manager_;
```
Comment (phrocker) [inline libminifi/include/ResourceClaim.h]: hmmm why are we creating a non-owning reference here? We could potentially lose data in some routes. Seems worrisome. I'll do some testing to validate.

File: libminifi/src/core/repository/VolatileContentRepository.cpp (hunk 1 above)
Comment (bakaid) [reply libminifi/src/core/repository/VolatileContentRepository.cpp]: As far as I could gather it served no discernible purpose, but because a shared_ptr instance is capture in a detached thread it causes valgrind to issue a warning (as valgrind will evaluate memory leaks at exit()).

File: libminifi/src/core/Processor.cpp (hunk 5 above)
Comment (bakaid) [reply libminifi/src/core/Processor.cpp]: I prefer nullptr too, but setDestination now gets a weak_ptr and weak_ptr has no constructor from nullptr_t (or any pointer for that matter).

**Comment by szaszm** [comment]
Closing this because of inactivity. Some of the leaks have been fixed on main, and I'll make sure to check this as reference when similar efforts are underway.
